pkg/resource/deploy/plan_executor.go

```diff
@@ -194,6 +194,10 @@ func (pe *planExecutor) Execute(callerCtx context.Context, opts Options, preview
 		err = execError("failed", preview)
 	} else if canceled {
 		err = execError("canceled", preview)
+	} else if pendingReplaces != nil {
+		// If execution completed successfully, we may have leftover resources that were pending replacment but never
+		// re-registered. We will record that these resources are no longer pending replacement here.
+		err = opts.Events.RemovePendingReplacements(pendingReplaces)
```

Comment:
> If execution completed successfully, we may have leftover resources that were pending replacment but never re-registered.

How can this happen?

Reply:
This is subtle.

Consider the following program:
```
const a = new Resource("a", { foo: "foo" });

const b = new Resource("b", { bar: a.out1 });

const c = new Resource("c", { baz: "baz" });
```

Let `Resource.foo` be a property that requires delete-before-replacement, `Resource.bar` be a property that requires replacement, and `baz` be a property that does not require replacement. 
Assume that we have run this program and all resources have been successfully created. Then, we change the inputs to `a` and `c`:

```
const a = new Resource("a", { foo: "bar" });

const b = new Resource("b", { bar: a.out1 });

const c = new Resource("c", { baz: "qux" });
```

Because `a` must be DBR'ed and a property from `a` flows into a reuqires-replacement property of `b`, both `b` and `a` should be deleted and replaced. `c` must be updated due to the change to `baz`, but this operation is independent of the others. When we run this update, we may see the following ordering:

1. delete `b` successfully
2. delete `a` successfully
3. recreate `a` successfully
4. attempt to update `c` and fail

The failure in step (4) causes the update to terminate before `b` is recreated. This leaves us with a `PendingReplacement` entry for `b` in the checkpoint file. Before the next update, we may change the program like so:

```
const a = new Resource("a", { foo: "bar" });

const c = new Resource("c", { baz: "qux" });
```

Assuming that the change to `c` now completes successfully, we will still have a `PendingReplacement` entry for `b` in the checkpoint. At this point, though, we know that `b` will never be re-registered, and can throw out its `PendingReplacement` entry.

Does that make sense?